[ASCollectionView][Architecture] Introduce ASSectionContext

nguyenhuy opened the pull request:
It is a custom object that can be optionally provided by ASCollectionDataSource and later retrieved from the collection view to be used in collection view layout. The idea was first proposed by @Adlai-Holler in #2067. 

I've been playing around with different ways to implement this feature, including having an ASSection that is a proxy of cell nodes array ([here](https://github.com/facebook/AsyncDisplayKit/compare/master...nguyenhuy:ASSectionContext_proxy?expand=1)), or an ASSection that is a subclass of NSMutableArray<ASCellNode *> ([here](https://github.com/facebook/AsyncDisplayKit/compare/master...nguyenhuy:ASSectionContext_array_subclass?expand=1)). I think the implementation in this PR is simpler and thus, more elegant.

Please let me know what you think, @Adlai-Holler @maicki @appleguy @garrettmoon @levi. Thanks!

TODO:
- [ ] Documentation ([Overview.md](https://github.com/facebook/AsyncDisplayKit/blob/master/plans/section-infos-api/Overview.md))
- [x] Unit test

Adlai-Holler commented on AsyncDisplayKit/ASCollectionView.h:
```diff
@@ -420,6 +423,8 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)collectionViewUnlockDataSource:(ASCollectionView *)collectionView ASDISPLAYNODE_DEPRECATED;
 
+- (_Nonnull id<ASSectionContext>)collectionView:(ASCollectionView *)collectionView contextForSectionAtIndex:(NSInteger)sectionIndex;
```
Shouldn't this be nullable?

nguyenhuy replied:
Exactly. This method is optional and if users implement it, they need to provide a context. 

That's said, in some cases, they may want to provide contexts for just a subset of sections. @Adlai-Holler Should we support it?

nguyenhuy replied:
Actually if we create an object for each section that holds a nullable context, we can relax on the nullability of this method. Will work on it.

Adlai-Holler commented on AsyncDisplayKit/Details/ASCollectionDataController.mm:
```diff
@@ -15,48 +15,67 @@
-  [_pendingContexts enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull kind, NSMutableArray<ASIndexedNodeContext *> * _Nonnull contexts, __unused BOOL * _Nonnull stop) {
+  NSIndexSet *sectionIndexSet = [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, newSectionCount)];
+  
+  [self applyPendingSectionContextsAtIndexSet:sectionIndexSet];
```
`applyPendingSectionContextsAtIndexes` and `_populateSectionContextsFromDataSourceForSections`

Adlai-Holler commented on AsyncDisplayKit/Details/ASCollectionDataController.mm:
```diff
@@ -271,6 +319,14 @@ - (ASCellNode *)supplementaryNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPat
+  ASDisplayNodeAssertMainThread();
+  ASDisplayNodeAssertTrue(_dataSourceSupportsSectionContext);
+  ASDisplayNodeAssertTrue(sectionIndex >= 0 && sectionIndex < _sectionContexts.count);
+  return _sectionContexts[sectionIndex];
```
I think we can be lenient and not assert that the data source supports section contexts here.

Adlai-Holler commented on AsyncDisplayKit.xcodeproj/project.pbxproj:
```diff
@@ -2039,6 +2046,21 @@
+			shellPath = /bin/sh;
+			shellScript = "\"${SRCROOT}/Pods/Target Support Files/Pods-AsyncDisplayKitTests/Pods-AsyncDisplayKitTests-frameworks.sh\"\n";
+			showEnvVarsInLog = 0;
+		};
```
Let's get this pods change out

Adlai-Holler commented on AsyncDisplayKit/Details/ASCollectionDataController.h:
```diff
@@ -42,4 +46,6 @@
 
 - (ASCellNode *)supplementaryNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath;
 
+- (id<ASSectionContext>)contextForSectionAtIndex:(NSInteger)sectionIndex;
```
`contextForSection:` in this and the data source method

Adlai-Holler commented:
@nguyenhuy This is looking perfect! This will be a huge boon for Pinterest, where synchronizing section info has been a major pain point.

Looking toward the future, I think creating an object for each section even if they don't provide us with a context is valuable. I have some pretty ambitious plans for collection view debugging tools, and reifying the section abstraction (with an auto-incrementing ID) will help move us toward that.

These section objects do not need to hold any nodes or anything – I agree with you that those other two approaches got too messy too fast. Just an ID and optionally a context. Thoughts?

Adlai-Holler commented:
@nguyenhuy How's this going? Let me know if there's something that seems wonky or not quite right – once landed this will resolve an entire class of possible inconsistency issues in one shot, so I'm really excited to get it dialed in.

nguyenhuy commented:
@Adlai-Holler: Sorry this took so long. I've added ASSection and addressed your comments. Please give this PR another review. Thank you!

appleguy commented:
@Adlai-Holler — is it OK if we merge this before @nguyenhuy 's final changes?  He's getting ready for the big move and so any polishing touches here would be better to do after integrating it, but only if they don't have important functional impacts.

nguyenhuy commented on AsyncDisplayKit/ASCollectionView.mm:
```diff
@@ -1013,6 +1022,21 @@ - (NSUInteger)dataController:(ASCollectionDataController *)dataController supple
+- (id<ASSectionContext>)dataController:(ASDataController *)dataController contextForSection:(NSInteger)section
+{
+  ASDisplayNodeAssertMainThread();
+  id<ASSectionContext> context;
```
Done

nguyenhuy commented on AsyncDisplayKit/ASCollectionView.h:
```diff
@@ -420,6 +423,8 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)collectionViewUnlockDataSource:(ASCollectionView *)collectionView ASDISPLAYNODE_DEPRECATED;
 
+- (id<ASSectionContext> _Nullable)collectionView:(ASCollectionView *)collectionView contextForSection:(NSInteger)section;
```
👌

nguyenhuy commented:
@Adlai-Holler @appleguy: Addressed comments and added test cases.

Adlai-Holler commented:
@nguyenhuy Wow excellent! I really appreciate the quality of your work. If you'll rebase or merge latest master, the CI should be fixed and this should be ready for primetime. 

I'm confident this will work well, but for the sake of due diligence we'll test this against Pinterest via a cherry-pick and then land it immediately after.

nguyenhuy commented:
@Adlai-Holler: Squashed all commits into one and rebased it on top of latest master. Waiting for Travis to finish which seems to take forever :(

Appreciate your test against Pinterest. Let me know what you find!

Adlai-Holler commented:
Hannah has tested it for a while and I did the same. This is running excellently! Thanks so much Huy – I'm going to start building on this system immediately. Let's do this again sometime!

nguyenhuy commented:
@Adlai-Holler: I'm really glad it works. More than happy to do this again, anytime!
